File: go/500.600/535.EncodeAndDecodeTinyurl.go
package main

import (
	"encoding/base64"
	"fmt"
	"strings"
)

type Codec struct {
}

func Constructor() Codec {
	return Codec{}
}

// Encodes a URL to a shortened URL.
func (this *Codec) encode(longUrl string) string {
	url := "http://tinyurl.com/" + base64.StdEncoding.EncodeToString([]byte(longUrl))
	return url
}

// Decodes a shortened URL to its original URL.
func (this *Codec) decode(shortUrl string) string {
	url := strings.TrimPrefix(shortUrl, "http://tinyurl.com/")
	byts, _ := base64.StdEncoding.DecodeString(url)
	return string(byts)
}

func main() {
	cls := Constructor()
	url := cls.encode("https://leetcode.com/problems/design-tinyurl")
	fmt.Println("Encode: ", url)
	fmt.Println("Decode", cls.decode(url))
}

File: 500.600/535.EncodeAndDecodeTinyurl.cpp
#include <iostream>
#include <string>
#include <unordered_map>
using namespace std;

class EncodeAndDecodeTinyurl
{
public:
    // Encodes a URL to a shortened URL.
    string encode(string longUrl)
    {
        int cur = ++index;
        string tiny = "http://tinyurl.com/" + to_string(cur);
        database[tiny] = longUrl;
        return tiny;
    }

    // Decodes a shortened URL to its original URL.
    string decode(string shortUrl)
    {
        return database[shortUrl];
    }

private:
    unordered_map<string, string> database;
    int index{};
};

int main()
{
    EncodeAndDecodeTinyurl cls;
    string url = cls.encode("https://leetcode.com/problems/design-tinyurl");
    cout << "Encode: " << url.c_str() << endl;
    cout << "Decode: " << cls.decode(url).c_str() << endl;
    return 0;
}
